### ISSUE TITLE
WebSocketProvider returns events for block `n - 1` when queried for block `n`

### ISSUE THREAD

**Issue by artbm** [mangrovedao/mangrove-ts#197]
**Description**

If I subscribe via WebSocketProvider, when mangrove emits an event e, I receive the event e-1, whereas when I subscribe via JsonRpcProvider, I receive the good event e. 

Here is an example : I post 2 bids. On the first bid, I receive the HTTP event. On the second bid, I receive the WebSocket previous event and the current HTTP event. You'll find the testing script at the end of this issue.

**To Reproduce**

1. Start your websocketHangIssue script
2. Post bid 1
    2.1. JsonRpc-based provider prints bid 1
3. Post bid 2
    3.1. JsonRpc-based provider prints bid 2
    3.2. WS-based provider prints bid 1

**Screenshots**

Points 1 => 2.1
![image](https://user-images.githubusercontent.com/24268913/153446030-66252c19-8e99-49c9-a0f0-fb40067f9310.png)
Points 3 => 3.2
![image](https://user-images.githubusercontent.com/24268913/153446288-6f7f215b-59ca-4573-90be-5f3b823bc76c.png)

**Envrionment**

 - OS: Ubuntu 20.04.3 LTS
 - Node v16.13.2
 - Yarn 3.1.0

**Code**

```js
const { Mangrove } = require("@mangrovedao/mangrove.js");
const { NonceManager } = require("@ethersproject/experimental");
const { Wallet } = require("@ethersproject/wallet");
const {
  WebSocketProvider,
  JsonRpcProvider,
} = require("@ethersproject/providers");
const chalk = require("chalk");

///////!\
const PRIVATE_KEY = "";
///////!\

const ETHEREUM_NODE_URL = "http://localhost:8545";

async function main() {
  listenHttp();
  listenWebsocket();
}

async function listenHttp() {
  const provider = new JsonRpcProvider(ETHEREUM_NODE_URL);
  const signer = new Wallet(PRIVATE_KEY, provider);
  const nonceManager = new NonceManager(signer);
  const mgv = await Mangrove.connect({ signer: nonceManager });
  const market = await mgv.market({ base: "WETH", quote: "DAI" });
  market.subscribe(async (event) => {
    console.log(
      chalk.green(
        `EVENT RECEIVED VIA HTTP : new ${event.type} ${event.ba} id:${event.offer.id}`
      )
    );
  });
}

async function listenWebsocket() {
  const provider = new WebSocketProvider(ETHEREUM_NODE_URL);
  const signer = new Wallet(PRIVATE_KEY, provider);
  const nonceManager = new NonceManager(signer);
  const mgv = await Mangrove.connect({ signer: nonceManager });
  const market = await mgv.market({ base: "WETH", quote: "DAI" });
  market.subscribe(async (event) => {
    console.log(
      chalk.red(
        `EVENT RECEIVED VIA WEBSOCKET : new ${event.type} ${event.ba} id:${event.offer.id}`
      )
    );
  });
}

main();
```

┆Issue is synchronized with this [Monday item](https://mangrovedao.monday.com/boards/3894180513/pulses/3901960572) by [Unito](https://www.unito.io)

**Comment by adhusson** [comment]
Note: adding @espendk so he takes a look, happy to investigate with you @espendk  :)

**Comment by espendk** [comment]
I think this could be related to issue #164  where we've discovered that ethers.js or hardhat sometimes reports the wrong logs.

**Comment by espendk** [comment]
Investigating #164, I've discovered that the problem is caused by `WebSocketProvider.getLogs` returning logs for the previous block instead of the block that is queried.

In particular, this affects the following code in `Semibook`: 

```javascript
      const logs = await this.market.mgv._provider.getLogs({
        fromBlock: blockNumber,
        toBlock: blockNumber,
        ...this.#eventFilter,
      });
```

When `getLogs` returns logs for block `blockNumber - 1` and no more blocks are mined, the events for block `blockNumber` are never processed.


I've reported the issue to ether.js: https://github.com/ethers-io/ethers.js/issues/3012


We've seen another issue which could also stem from this: #205.
Once this issue has been fixed, we should check if #205 has been fixed as well.

**Comment by dontrolle** [comment]
Since we're waiting for an external signal, I'll desassign myself, as I'm not going to be working on it at this moment.

**Comment by peterMangrove** [comment]
@dontrolle Is this solved by the #953 issue that you solved?

**Comment by dontrolle** [comment]
I've removed the code in question so I'm
